menas/ui/components/dataset/conformanceRule/upsert.controller.js

```diff
@@ -0,0 +1,265 @@
+      let toks = sBindPath.split("/");
+      let inputColumnIndex = parseInt(toks[toks.length - 1]);
+      let oldInputColumns = this._model.getProperty("/newRule/inputColumns");
+
+      let newInputColumns = oldInputColumns.filter((_, index) => index !== inputColumnIndex);
+      this._model.setProperty("/newRule/inputColumns", newInputColumns);
+    },
+
+    onDeleteJoinCondition: function (oEv) {
+      let sBindPath = oEv.getParameter("listItem").getBindingContext().getPath();
+      let toks = sBindPath.split("/");
+      let inputColumnIndex = parseInt(toks[toks.length - 1]);
+      let oldInputColumns = this._model.getProperty("/newRule/newJoinConditions");
+
+      let newInputColumns = oldInputColumns.filter((_, index) => index !== inputColumnIndex);
+      this._model.setProperty("/newRule/newJoinConditions", newInputColumns);
+    },
+
+    onSchemaFieldSelect: function (oEv) {
+      let bind = oEv.getParameter("listItem").getBindingContext().getPath();
+      let modelPathBase = "/currentDataset/schema/fields/";
+      let model = sap.ui.getCore().getModel();
+      SchemaService.fieldSelect(bind, modelPathBase, model, "/newRule/inputColumn");
+    },
+
+    beforeShowFragmentChanges: function () {
+      let currentRule = this._model.getProperty("/newRule");
+      let newRule = currentRule;
+
+      if (currentRule._t === "ConcatenationConformanceRule" && !currentRule.isEdit) {
+        newRule.inputColumns = ["", ""];
+      }
+
+      if (currentRule._t === "MappingConformanceRule") {
+        if(!currentRule.isEdit){
+          newRule.newJoinConditions = [{mappingTableField: "", datasetField: ""}];
+          newRule.mappingTable = this._model.getProperty("/mappingTables")[0]._id;
+        } else {
+          let oAttributeMappings = newRule.attributeMappings;
+          let aNewJoinConditions = [];
+          for (let key in oAttributeMappings) {
+            aNewJoinConditions.push({
+              mappingTableField: key,
+              datasetField: oAttributeMappings[key]
+            });
+          }
+          newRule.newJoinConditions = aNewJoinConditions;
+        }
+        MappingTableService.getAllMappingTableVersions(newRule.mappingTable, sap.ui.getCore().byId("schemaVersionSelect"));
+      }
+
+      if(!currentRule.isEdit)
+        newRule.order = this._model.getProperty("/currentDataset").conformance.length;
+
+      this._model.setProperty("/newRule", newRule);
+    },
+
+    beforeSubmitChanges: function (newRule) {
+      if (newRule._t === "MappingConformanceRule") {
+        newRule.attributeMappings = {};
+        newRule.newJoinConditions.map(function(joinCondition) {
+          newRule.attributeMappings[joinCondition.mappingTableField] = joinCondition.datasetField
+        });
+        delete newRule.joinConditions;
+      }
+    },
+
+    addRule: function (currentDataset, newRule) {
+      currentDataset.conformance[newRule.order] = newRule;
```

Comment:
Pretty sure `updateRule` and `addRule` are the same function. It makes sense to have the logical separation fo the two functions, so we can use it to distinguish the two intended behaviours. This distinction will allow us to more easily refactor further, if we decide to slim down the front-end (I suggested we fatten it, but not so sure about that anymore).
The only change I would make here is, for now, to call `updateRule` from `addRule`, so when we change the front-end logic for `addRule` nothing else is affected.

Reply:
Fixed.